test/utils/map-state.spec.js

```diff
@@ -0,0 +1,184 @@
+import test from 'tape-catch';
+import MapState from '../../src/utils/map-state';
+import {PerspectiveMercatorViewport} from 'viewport-mercator-project';
+
+const SAMPLE_VIEWPORTS = [
+  // SF
+  {
+    width: 800,
+    height: 600,
+    longitude: -122.58,
+    latitude: 37.74,
+    zoom: 14
+  },
+  // Edge location, custom zoom limit
+  {
+    width: 800,
+    height: 600,
+    longitude: -179,
+    latitude: 90,
+    zoom: 0,
+    maxZoom: 0.5
+  },
+  // SF with rotation, custom pitch limit
+  {
+    width: 800,
+    height: 600,
+    longitude: -122.58,
+    latitude: 37.74,
+    zoom: 14,
+    pitch: 60,
+    bearing: 45,
+    maxPitch: 90
+  }
+];
+
+// Discard precision errors for comparison
+function toLowPrecision(input, precision = 11) {
+  if (typeof input === 'number') {
```

Comment:
Nit: Maybe make this a one-line ternary? Making this shorter will detract less from the tests.